### PR TITLE
tools/rados: always call rados.shutdown() before exit()

### PR DESCRIPTION
**Pull request by dragonylffly**
When connected to ceph cluster, if call exit(1) directly, will cause the finisher
thread segmentation fault as follows,

Caught signal (Segmentation fault) **in thread 7f45377fe700 thread_name:fn_anonymous

run 'rados mapext' without other arguments could easily get it. This patch fixes it
by calling rados.shutdown() before exit()

Fixes: http://tracker.ceph.com/issues/36732

Signed-off-by: Li Wang <laurence.liwang@gmail.com>

### CODE REVIEW

File: src/tools/rados/rados.cc [hunk 1]
```diff
@@ -2116,6 +2116,7 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
 
   if (create_pool && !pool_name) {
     cerr << "--create-pool requested but pool_name was not specified!" << std::endl;
+    rados.shutdown();
```
Comment (tchaikov) [inline src/tools/rados/rados.cc]: i think an alternative approach is to instead call usage(cerr) and return `1`, and leave the cleanup to the destructor of `rados`.

**Comment by tchaikov** [comment]
also, the title of `fix segmentation fault` is a little bit vague, better off using something like 

> always call rados.shutdown() before exit()

and explain the rationale of this change.

File: src/tools/rados/rados.cc (hunk 1 above)
Comment (dragonylffly) [reply src/tools/rados/rados.cc]: i also considered this, and thought either way is ok. If you insist, i could revise it

**Comment by dragonylffly** [comment]
@tchaikov updated

**Comment by tchaikov** [comment]
@dragonylffly ping?

**Comment by dragonylffly** [comment]
will submit a new version

**Comment by dragonylffly** [comment]
@tchaikov updated

File: src/tools/rados/rados.cc [hunk 2]
```diff
@@ -2918,113 +2984,25 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
...
+      goto out;
+    }
     string oid(nargs[1]);
     ret = io_ctx.create(oid, true);
     if (ret < 0) {
       cerr << "error creating " << pool_name << "/" << oid << ": " << cpp_strerror(ret) << std::endl;
       goto out;
     }
   }
-
```
Comment (tchaikov) [inline src/tools/rados/rados.cc]: this change does not belong to this PR. could you rebase onto master and repush?

File: src/tools/rados/rados.cc [hunk 3]
```diff
@@ -2896,8 +2959,11 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
       goto out;
     }
   } else if (strcmp(nargs[0], "rm") == 0) {
-    if (!pool_name || nargs.size() < 2)
-      usage_exit();
+    if (!pool_name || nargs.size() < 2) {
+      usage(cerr);
+      ret = -1;
+      goto out;
```
Comment (tchaikov) [inline src/tools/rados/rados.cc]: the whole point of `goto out` is to `delete formatter`. but guess we can fix this by managing it using a unique_ptr<> in a follow-up PR.

Comment (dragonylffly) [reply src/tools/rados/rados.cc]: @tchaikov updated with unique_ptr used

**Comment by dragonylffly** [comment]
@tchaikov udpated with unique_ptr used

File: src/tools/rados/rados.cc [hunk 4]
```diff
@@ -2034,7 +2035,8 @@ static int rados_tool_common(const std::map < std::string, std::string > &opts,
   i = opts.find("format");
   if (i != opts.end()) {
     const char *format = i->second.c_str();
-    formatter = Formatter::create(format);
+    _formatter.reset(Formatter::create(format));
+    formatter = _formatter.get();
```
Comment (tchaikov) [inline src/tools/rados/rados.cc]: we assign to `formatter` at multiple places, like
```c++
    if (!formatter) {
      formatter = new JSONFormatter(pretty_format);
    }
```
so this change leaks.

instead, i'd suggest put just use the unique_ptr when referencing the Formatter, instead of using the raw pointer.

and update the places like above to something like
```c++
    if (!formatter) {
      formatter = std::make_unique<JSONFormatter>(pretty_format);
    }
```

Comment (dragonylffly) [reply src/tools/rados/rados.cc]: those places revised already, and some functions take raw pointer as input, so we have to leave the raw pointer, otherwise not compile

**Comment by dragonylffly** [comment]
@tchaikov updated, btw, please take a look at: #25128

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-11-28_00:59:04-rados-wip-kefu2-testing-2018-11-27-2345-distro-basic-smithi/

failures tracked by / caused by
- http://tracker.ceph.com/issues/37432
- http://tracker.ceph.com/issues/20798
- http://tracker.ceph.com/issues/37391
- #19454
